### ISSUE TITLE
Some questions

### ISSUE THREAD

**Issue by julianomqs** [vriad/zod#142]
Hello,

I'm giving a try of your lib in my project, and it looks quite nice.

I have some questions, sorry if this is not the best place to do it, but I didn't find a tag in stackoverflow...

1 - Can I use this lib in the backend and frontend, or just in one of two?

2 - I'm trying to do the following rule:
The **bank** property (an object) of an object (a React form) is required.
But I don't care about **bank** properties, just if it is informed (not undefined) or not, because it comes from an autocomplete component.

I'm doing like this:

```typescript
z.object({
  bank: z
    .object({
      id: z.union([z.number(), z.string()])
    })
.nonstrict(),
```

Is this the best way?

3 - How to extend this lib with custom methods?
For example, a string validation method to validate CPF (a SSN like number in my country).
I did a refine and it worked well, but I'm thinking about extending the schema, like this:

```typescript
z.string().cpf()
```

Can I just reexport zod with my custom methods like this:

```typescript
const CustomZod = {
  string: () =>
    z.string().refine(
      arg => validateCPF(arg),
      { message: "CPF inválido" }
    )
};
```

Is this the best way?

4 - I'm using Formik, and in Formik the validation function receives an object of values and must return an object with the error messages, one message for each key. I'm doing like this:

```typescript
const validateZod = (values: unknown): object => {
    const parsingResult = props.validationSchema.safeParse(values);

    if (parsingResult.success) {
      return {};
    }

    const flattenedErrors = parsingResult.error.flatten();

    const errors = {};

    for (const [key, value] of Object.entries(flattenedErrors.fieldErrors)) {
      errors[key] = value[0];
    }

    return errors;
  };
```

Is this the best way?

5 - I want to add a functionality to add asterisks to required fields in Formik.
So basically I would need to read the Zod schema passed to form, and check what fields of schema are required to put the asterisk.
Basically a reverse engineering of sorts.
How can I do that with Zod?

Thanks a lot.

**Comment by vriad** [comment]
Hey Juliano,

Good questions:

> 1 - Can I use this lib in the backend and frontend, or just in one of two?
You can use Zod in both. I added this to the readme as well. 

> 2 - I'm trying to do the following rule:
> The bank property (an object) of an object (a React form) is required.
> But I don't care about bank properties, just if it is informed (not undefined) or not, because it comes from an autocomplete component.
> 
> I'm doing like this:
> 
> z.object({
>   bank: z
>     .object({
>       id: z.union([z.number(), z.string()])
>     })
> .nonstrict(),
> Is this the best way?

Yep, that's the solution I'd recommend. You don't even need to include a 

> 3 - How to extend this lib with custom methods?
> For example, a string validation method to validate CPF (a SSN like number in my country).
> I did a refine and it worked well, but I'm thinking about extending the schema, like this:
> 
> z.string().cpf()
> Can I just reexport zod with my custom methods like this:
> 
> const CustomZod = {
>   string: () =>
>     z.string().refine(
>       arg => validateCPF(arg),
>       { message: "CPF inválido" }
>     )
> };
> Is this the best way?

That code won't let you do do `z.string().cpf()` but it's still a good approach. You could re-export all of zod if you want a more "native" feel:

```ts
// customzod.ts
export * from "zod";

export const cpf = ()=>z.string().refine(
  arg => validateCPF(arg),
  { message: "CPF inválido" }
);


// index.ts
import * as z from "./customzod";
const cpfSchema = z.cpf();
```

To achieve `z.string().cpf()` you'd have to subclass ZodString:

```ts
import * as z from 'zod';

export class MyZodString extends z.ZodString {
  cpf = () => this.refine(arg => validateCPF(arg), { message: 'CPF inválido' });
}
```

> 4 - I'm using Formik, and in Formik the validation function receives an object of values and must return an object with the error messages, one message for each key. I'm doing like this:
> 
> const validateZod = (values: unknown): object => {
>     const parsingResult = props.validationSchema.safeParse(values);
> 
>     if (parsingResult.success) {
>       return {};
>     }
> 
>     const flattenedErrors = parsingResult.error.flatten();
> 
>     const errors = {};
> 
>     for (const [key, value] of Object.entries(flattenedErrors.fieldErrors)) {
>       errors[key] = value[0];
>     }
> 
>     return errors;
>   };
> Is this the best way?


This seems like a good solution as well though it will only provide the first error associated with each key. Depending on how formik handles newlines, you may want to  consider this:

```ts
errors[key] = value.join('\n');
```

> 5 - I want to add a functionality to add asterisks to required fields in Formik.
> So basically I would need to read the Zod schema passed to form, and check what fields of schema are required to put the asterisk.
> Basically a reverse engineering of sorts.
> How can I do that with Zod?

Here's a utility function for determining if a given schema is "optional":

```ts
import * as z from 'zod';

export const isOptional = <T extends z.ZodTypeAny>(schema: T) => {
  if (
    schema instanceof z.ZodUndefined
  ) {
    return true;
  } else if (schema instanceof z.ZodUnion) {
    for (const opt of schema._def.options) {
      if (isOptional(opt)) {
        return true;
      }
    }
    return false;
  } else {
    return false;
  }
};
```

You can loop over your form schema's `.shape` property (assuming your form schema is an object schema) like so:

```ts
const optionals: object = {};
for(const key in myForm.shape){
  optionals[key] = isOptional(myForm.shape[key]);
}
```

Not tested but it'll probably work. 

Good questions 👍

**Comment by julianomqs** [comment]
Thanks a lot.

**Comment by tmtron** [comment]
This looks almost like the [isOptiona()](https://github.com/vriad/zod/blob/1.11/src/codegen.ts#L6) function in `codegen.ts`. 

@vriad What do you think of exporting this function, so that we don't need to copy the code?

**Comment by vriad** [comment]
Will do, probably in Zod 2 👍

**Comment by vriad** [comment]
@julianomqs @tmtron just thought of a better `isOptional` implementation!

```ts
const isOptional = (schema: z.ZodType<any, any>): boolean => schema.safeParse(undefined).success;
```

So might be not be worth putting into core after all. 😛
